lib/puppet/provider/package/dpkg.rb

```diff
@@ -113,6 +120,20 @@ def query
 
     # list out our specific package
     begin
+      if @resource.allow_virtual?
+        output = dpkgquery(
+          "-W",
+          "--showformat",
+          self.class::DPKG_QUERY_PROVIDES_FORMAT_STRING
+          #the regex searches for the resource[:name] in the dpkquery result in which the Provides field is also available
+          #it will search for the packages only in the brackets ex: [rubygems]
+        ).lines.find {|package| package.match(/[\[ ](#{Regexp.escape(@resource[:name])})[\],]/)}
+        if output          
+          hash = self.class.parse_line(output,self.class::FIELDS_REGEX_WITH_PROVIDES)
+          Puppet.info("Package #{@resource[:name]} is virtual, defaulting to #{hash[:name]}")
```

Comment:
Can `hash` be nil?

Reply:
In this case it can't be because the output matches the exact pattern that can be extracted with parse_line